File: scm-csharp/scheme/SchemeConditionVariable.cs
using System.Threading;

namespace scheme;

public class SchemeConditionVariable
{
    public readonly object Lock = new object();
    public object name = Value.F;
    public object specific = Value.NIL;

    public void Signal()
    {
        lock (Lock) { Monitor.Pulse(Lock); }
    }

    public void Broadcast()
    {
        lock (Lock) { Monitor.PulseAll(Lock); }
    }

    public bool Wait(int timeoutMs)
    {
        lock (Lock) { return Monitor.Wait(Lock, timeoutMs); }
    }

    public void Wait()
    {
        lock (Lock) { Monitor.Wait(Lock); }
    }

    public override string ToString() =>
        !name.Equals(Value.F) ? "#<condition-variable " + Value.DisplayRep(name) + ">" : "#<condition-variable>";
}

File: scm-java/src/main/scheme/SchemeConditionVariable.java
package scheme;

import java.util.concurrent.locks.ReentrantLock;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.TimeUnit;

public class SchemeConditionVariable {
    public final ReentrantLock lock = new ReentrantLock();
    public final Condition condition = lock.newCondition();
    public Object name = Value.F;
    public Object specific = Value.NIL;

    public void signal() {
        lock.lock();
        try { condition.signal(); }
        finally { lock.unlock(); }
    }

    public void broadcast() {
        lock.lock();
        try { condition.signalAll(); }
        finally { lock.unlock(); }
    }

    public boolean await(long timeoutMs) throws InterruptedException {
        lock.lock();
        try { return condition.await(timeoutMs, TimeUnit.MILLISECONDS); }
        finally { lock.unlock(); }
    }

    public void awaitUninterruptibly() {
        lock.lock();
        try { condition.await(); }
        catch (InterruptedException e) { Thread.currentThread().interrupt(); }
        finally { lock.unlock(); }
    }

    @Override
    public String toString() {
        return name != Value.F ? "#<condition-variable " + Value.displayRep(name) + ">" : "#<condition-variable>";
    }
}
